src/db/collections/block.rs

```diff
@@ -1,65 +1,84 @@
+use futures::TryStreamExt;
 use mongodb::{
     bson::{self, doc},
     error::Error,
-    options::UpdateOptions,
+    options::{FindOneOptions, IndexOptions, UpdateOptions},
+    IndexModel,
 };
 use serde::{Deserialize, Serialize};
 
 use crate::{
     db::MongoDb,
     types::{
-        ledger::{BlockMetadata, LedgerInclusionState},
-        stardust::block::{Address, Block, BlockId, TransactionId},
-        tangle::MilestoneIndex,
+        ledger::{BlockMetadata, LedgerInclusionState, OutputMetadata, OutputWithMetadata, SpentMetadata},
+        stardust::block::{Block, BlockId, Output, OutputId, TransactionId},
     },
 };
 
 /// Chronicle Block record.
 #[derive(Clone, Debug, Serialize, Deserialize)]
-struct BlockDocument {
+pub struct BlockDocument {
     /// The id of the current block.
-    block_id: BlockId,
+    pub block_id: BlockId,
     /// The block.
-    block: Block,
+    pub block: Block,
     /// The raw bytes of the block.
     #[serde(with = "serde_bytes")]
-    raw: Vec<u8>,
+    pub raw: Vec<u8>,
     /// The block's metadata.
-    metadata: BlockMetadata,
+    pub metadata: BlockMetadata,
     /// The index of this block in white flag order.
-    white_flag_index: u32,
+    pub white_flag_index: u32,
 }
 
 impl BlockDocument {
     /// The stardust blocks collection name.
     const COLLECTION: &'static str = "stardust_blocks";
 }
 
-/// A single transaction history result row.
-#[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct TransactionHistoryResult {
-    /// The transaction id.
-    pub transaction_id: TransactionId,
-    /// The index of the output that this transfer represents.
-    pub output_index: u16,
-    /// Whether this is a spent or unspent output.
-    pub is_spent: bool,
-    /// The inclusion state of the output's transaction.
-    pub inclusion_state: Option<LedgerInclusionState>,
-    /// The transaction's block id.
-    pub block_id: BlockId,
-    /// The milestone index that references the transaction.
-    pub milestone_index: Option<MilestoneIndex>,
-    /// The transfer amount.
-    pub amount: u64,
-}
-
 /// Implements the queries for the core API.
 impl MongoDb {
+    /// Creates block indexes.
+    pub async fn create_block_indexes(&self) -> Result<(), Error> {
+        let collection = self.0.collection::<BlockDocument>(BlockDocument::COLLECTION);
+
+        collection
+            .create_index(
+                IndexModel::builder()
+                    .keys(doc! { "block_id": 1 })
+                    .options(
+                        IndexOptions::builder()
+                            .unique(true)
+                            .name("block_id_index".to_string())
+                            .build(),
+                    )
+                    .build(),
+                None,
+            )
+            .await?;
+
+        collection
+            .create_index(
+                IndexModel::builder()
+                    .keys(doc! { "block.payload.transaction_id": 1 })
```

Comment:
is it not "block.payload.transaction.id" ?
(`Block` > `Payload` > `TransactionPayload` > `TransactionId`)